File: js/Connection.js
// Wrap a given WebSocket connection with the asyncProtocol
// Events: error(err), call(type, data, answer), close()
function Connection(webSocket) {
	// Store the underlying webSocket
	this.webSocket = webSocket
	
	// Store the last received and sent auto-increment ids
	this._lastReceivedID = 0
	this._lastSentID = 0
	
	// List of calls waiting for an answer
	this._calls = []
	
	// Set listeners for webSocket events
	this.webSocket.that = this
	this.webSocket.onclose = this._onclose
	this.webSocket.onerror = this._onerror
	this.webSocket.onmessage = this._processMessage
}

// Register a new type of call that the server can make
Connection.registerServerCall = function (id, argsFormat, returnFormat) {
	if (Math.round(id) != id || id < 1)
		throw new TypeError("id must be a non-zero unsigned integer")
	if (id in Connection._registeredServerCalls)
		throw new Error("Unable to register server call "+id+", it has already been registered")
	Connection._registeredServerCalls[id] = [inflateFormat(argsFormat), inflateFormat(returnFormat)]
	return id
}

// Register a new type of call that clients can make
Connection.registerClientCall = function (id, argsFormat, returnFormat) {
	if (Math.round(id) != id || id < 1)
		throw new TypeError("id must be a non-zero unsigned integer")
	if (id in Connection._registeredClientCalls)
		throw new Error("Unable to register client call "+id+", it has already been registered")
	Connection._registeredClientCalls[id] = [inflateFormat(argsFormat), inflateFormat(returnFormat)]
	return id
}

// Register a new type of exception
Connection.registerException = function (id, dataFormat) {
	if (Math.round(id) != id || id < 1)
		throw new TypeError("id must be a non-zero unsigned integer")
	if (id in Connection._registeredExceptions)
		throw new Error("Unable to register exception "+id+", it has already been registered")
	Connection._registeredExceptions[id] = inflateFormat(dataFormat)
	return id
}

// Send a call to the other side
// type is the call-type id (int)
// data is the argument data (optional, must be a Data, DataArray or string. Must match the format registered with Connection.registerClientCall)
// onreturn(data) is a callback (optional)
// onexception(type, data) is a callback (optional)
// timeout is the maximum time this endpoint will wait for a return/exception (optional, default: 60e3)
// Inside the callbacks, this will be the Connection object
Connection.prototype.sendCall = function (type, data, onreturn, onexception, timeout) {
	var meta, length, interval, call
	
	// Validates the data
	call = Connection._registeredClientCalls[type]
	if (!call)
		throw new Error("Invalid call type "+type)
	data = Data.toData(data)
	if (data.format != call[0].formatString)
		throw new Error("Invalid data type '"+data.format+"' for call "+type)
	
	// Creates the protocol meta-data
	data = data.toBuffer()
	meta = (new Data).addUint(type).addUint(++this._lastSentID).toBuffer()
	length = (new Data).addUint(meta.length+data.length).toBuffer()
	
	// Send the message
	this.webSocket.send(new Uint8Array([length, meta, data]).buffer)
	
	// Set timeout
	timeout = timeout===undefined ? 60e3 : timeout
	if (timeout)
		interval = setTimeout(this._getTimeoutCallback(), timeout)
	
	// Save info about the sent call
	// [expectedReturnFormat, onreturn, onexception, interval]
	this._calls[this._lastSentID] = [call[1], onreturn, onexception, interval]
}

// Close the connection
Connection.prototype.close = function () {
	this.webSocket.close()
}

// Registered calls
Connection._registeredServerCalls = {}
Connection._registeredClientCalls = {}
Connection._registeredExceptions = {}

// Returns a callback to treat the timeout
Connection.prototype._getTimeoutCallback = function () {
	var that = this, id = this._lastSentID
	return function () {
		call = that._calls[id]
		if (call) {
			if (call[3])
				clearInterval(call[3])
			if (call[2])
				call[2].call(that, 0, null)
			delete that._calls[id]
		}
	}
}

// Let the connection be closed
Connection.prototype._onerror = function (err) {
}

// Inform the connection has been closed (send -1 exception to every pending call)
Connection.prototype._onclose = function () {
	var i, call, that
	that = this.that
	that.emit("close")
	for (i in that._calls)
		// Foreach openned call, dispatch the error exception
		if (that._calls.hasOwnProperty(i)) {
			call = that._calls[i]
			if (call[3])
				clearInterval(call[3])
			if (call[2])
				call[2].call(that, -1, null)
		}
	
	// Clear everything
	that._calls = {}
}

// Process the incoming message (a MessageEvent)
Connection.prototype._processMessage = function (message) {
	var aux, type, callID, offset
	
	// Extracts the message type and sequence id
	message = new Uint8Array(message.data)
	try {
		aux = []
		offset = inflateData.readUint(message, 0, aux)
		offset = inflateData.readUint(message, offset, aux)
		type = aux[0]
		callID = aux[1]
	
		if (type)
			// A call from the other side
			this._processCall(callID, type, message.subarray(offset))
		else {
			offset = inflateData.readUint(message, offset, aux)
			type = aux[2]
			if (type)
				// An exception from the other side
				this._processException(callID, type, message.subarray(offset))
			else
				// A return from the other side
				this._processReturn(callID, message.subarray(offset))
		}
	} catch (e) {
		this._protocolError()
	}
}

// Process an incoming call
Connection.prototype._processCall = function (callID, type, dataBuffer) {
	var call, data, answer, answered, that = this
	
	// Get call definition
	call = Connection._registeredServerCalls[type]
	if (!call) {
		this._protocolError()
		return
	}
	
	// Read the incoming data
	try {
		data = inflateData(dataBuffer, call[0])
	} catch (e) {
		// Invalid format
		this._protocolError()
		return
	}
	
	// Create the answer callback
	// obj can be an Exception or a Data (or convertable to Data) in the call-return format
	answered = false
	answer = function (obj) {
		var data
		if (answered)
			throw new Error("Answer already sent")
		if (obj instanceof Exception)
			that._sendAnswer(callID, obj.type, obj.data)
		else {
			data = Data.toData(obj)
			if (data.format != call[1].formatString)
				throw new Error("Invalid data type '"+data.format+"' for return "+type)
			that._sendAnswer(callID, 0, data)
		}
		answered = true
	}
	
	// Emmits the "call" event
	this.emit("call", type, data, answer)
}

// Process a return
Connection.prototype._processReturn = function (callID, dataBuffer) {
	var callInfo, data
	
	callInfo = this._calls[callID]
	if (!callInfo) {
		// Received a timeouted (or invalid) answer
		this._protocolError()
		return
	}
	
	// Read the incoming data
	try {
		data = inflateData(dataBuffer, callInfo[0])
	} catch (e) {
		// Invalid format
		this._protocolError()
		return
	}
	
	// Clear the timeout
	if (callInfo[3])
		clearInterval(callInfo[3])
	
	// Call the callback
	if (callInfo[1])
		callInfo[1].call(this, data)
	delete this._calls[callID]
}

// Process a returned exception
Connection.prototype._processException = function (callID, type, dataBuffer) {
	var callInfo, data, format
	
	callInfo = this._calls[callID]
	if (!callInfo) {
		// Received a timeouted (or invalid) answer
		this._protocolError()
		return
	}
	
	// Get exception definition
	format = Connection._registeredExceptions[type]
	if (!format) {
		this._protocolError()
		return
	}
	
	// Read the incoming data
	try {
		data = inflateData(dataBuffer, format)
	} catch (e) {
		// Invalid format
		this._protocolError()
		return
	}
	
	// Clear the timeout
	if (callInfo[3])
		clearInterval(callInfo[3])
	
	// Call the callback
	if (callInfo[2])
		callInfo[2].call(this, type, data)
	delete this._calls[callID]
}

// Treats a protocol error (close the connection)
Connection.prototype._protocolError = function () {
	this.webSocket.close(1002)
}

// Sends an answer (return or exception)
Connection.prototype._sendAnswer = function (callID, exceptionType, data) {
	var meta, length
	
	// Creates the buffers
	data = data.toBuffer()
	meta = (new Data).addUint(0).addUint(callID).addUint(exceptionType).toBuffer()
	length = (new Data).addUint(meta.length+data.length).toBuffer()
	
	// Send the message
	this.webSocket.send(new Uint8Array([length, meta, data]).buffer)
}

File: c#/AsyncProtocol/Connection.cs
using System;
using System.Collections.Generic;
using System.Timers;

namespace Sitegui.AsyncProtocol {
	/// <summary>
	/// Represent an async-protocol connection
	/// </summary>
	public class Connection : IDisposable {
		/// <summary>
		/// Represent the object sent to "call" event
		/// </summary>
		public class CallEventArgs : EventArgs {
			/// <summary>
			/// The call type
			/// </summary>
			public uint Type { get { return Call.Id; } }

			/// <summary>
			/// The inflated data
			/// </summary>
			public readonly object Data;

			bool answered = false;
			Connection Conn;
			Registry.RegisteredCall Call;
			uint CallId;

			/// <summary>
			/// Create a new args bag
			/// </summary>
			/// <param name="type">The type of this call</param>
			/// <param name="data">The inflated data</param>
			/// <param name="conn">The connection that created this bag</param>
			/// <param name="call">The call definition</param>
			/// <param name="callId">The call sequence id</param>
			internal CallEventArgs(Registry.RegisteredCall call, object data, Connection conn, uint callId) {
				Data = data;
				Conn = conn;
				Call = call;
				CallId = callId;
			}

			/// <summary>
			/// Answer the call with an exception
			/// </summary>
			/// <param name="data">The exception to be sent</param>
			/// <returns>Return whether the answer was sent (false if the connection is closed)</returns>
			public bool Answer(CallException data) {
				return Answer(data.Type, data.Data);
			}

			/// <summary>
			/// Answer the call with an empty return
			/// </summary>
			/// <returns>Return whether the answer was sent (false if the connection is closed)</returns>
			public bool Answer() {
				return Answer(0, new Data());
			}

			/// <summary>
			/// Answer the call with a given return
			/// </summary>
			/// <param name="data">The data pack to send</param>
			/// <returns>Return whether the answer was sent (false if the connection is closed)</returns>
			public bool Answer(Data data) {
				return Answer(0, data);
			}

			/// <summary>
			/// Answer the call with a string
			/// </summary>
			/// <param name="data">The string to send</param>
			/// <returns>Return whether the answer was sent (false if the connection is closed)</returns>
			public bool Answer(string data) {
				return Answer(0, new Data(data));
			}

			/// <summary>
			/// Internal method to send the answer
			/// </summary>
			/// <param name="exceptionType">The exception type (0 means no exception, normal return instead)</param>
			/// <param name="data">The data pack to send</param>
			/// <returns>Return whether the answer was sent (false if the connection is closed)</returns>
			bool Answer(uint exceptionType, Data data) {
				// Validate the connection
				if (answered)
					throw new InvalidOperationException("Answer already sent");
				if (!Conn.IsReady)
					return false;
				
				// Validate the arguments
				if (exceptionType != 0) {
					if (!Call.HasException(exceptionType))
						throw new ArgumentException("Invalid exception " + exceptionType + " to call " + Call.Id);
				} else {
					if (data.Format != Call.ReturnFormat.FormatString)
						throw new ArgumentException("Invalid data type '" + data.Format + "' for return type " + Call.Id);
				}

				// Send the answer
				byte[] binData = data.GetBytes();
				byte[] binMeta = new Data().AddUint((ulong)0).AddInt(CallId).AddInt(exceptionType).GetBytes();
				byte[] binLength = new Data().AddUint((ulong)(binData.Length + binMeta.Length)).GetBytes();
				Conn.Socket.Write(binLength);
				Conn.Socket.Write(binMeta);
				Conn.Socket.Write(binData);
				answered = true;
				return true;
			}
		}

		/// <summary>
		/// Represent a client call waiting for an answer from the server
		/// </summary>
		class PendingCall {
			public uint CallId;
			public Registry.RegisteredCall Call;
			public ReturnDelegate OnReturn;
			public ExceptionDelegate OnException;
			public Timer Interval;

			public PendingCall(uint callId, Registry.RegisteredCall call, ReturnDelegate onReturn, ExceptionDelegate onException, Timer interval) {
				CallId = callId;
				Call = call;
				OnReturn = onReturn;
				OnException = onException;
				Interval = interval;
			}
		}

		/// <summary>
		/// Fired when the connection is ready
		/// </summary>
		public event EventHandler OnOpen;

		/// <summary>
		/// Fired when a server call arrives
		/// </summary>
		public event EventHandler<CallEventArgs> OnCall;

		/// <summary>
		/// Fired when the connection is closed (normally or due an protocol error)
		/// </summary>
		public event EventHandler OnClose;

		/// <summary>
		/// Fired when something went wrong.
		/// The connection will be closed if opened.
		/// If there is no registered listener to this event, the exeception will be rethrow
		/// </summary>
		public event EventHandler<EventSocket.ExceptionEventArgs> OnError;

		/// <summary>
		/// The delegate for "return" callback in Send()
		/// </summary>
		/// <param name="sender">This connection itself</param>
		/// <param name="data">The inflated data returned by the server</param>
		public delegate void ReturnDelegate(Connection sender, object data);

		/// <summary>
		/// The delegate for "exception" callback in Send()
		/// </summary>
		/// <param name="sender">This connection itself</param>
		/// <param name="type">The type of exception (0 means the call has timeouted, -1 means the connection was closed)</param>
		/// <param name="data">The inflated data returned by the server (null for type 0 and -1)</param>
		public delegate void ExceptionDelegate(Connection sender, int type, object data);

		/// <summary>
		/// Indicate whether this connection is ready to send and receive data
		/// </summary>
		public bool IsReady { get { return Socket != null && Socket.IsReady; } }

		EventSocket Socket;
		uint LastReceivedId;
		uint LastSentId;
		BufferView Cache = new BufferView();
		readonly LinkedList<PendingCall> PendingCalls = new LinkedList<PendingCall>();
		bool disposed = false;

		/// <summary>
		/// Create a new connection and start connecting
		/// </summary>
		/// <param name="host">The host to connect to</param>
		/// <param name="port">The port to use</param>
		public Connection(string host, int port) {
			Socket = new EventSocket(host, port);
			Socket.OnData += Socket_OnData;
			Socket.OnError += Socket_OnError;
			Socket.OnClose += Socket_OnClose;
			Socket.OnConnect += Socket_OnConnect;
			LastReceivedId = 0;
			LastSentId = 0;
		}

		/// <summary>
		/// Dispose resources. You don't need to call this directly, instead use End()
		/// </summary>
		public void Dispose() {
			Dispose(true);
			GC.SuppressFinalize(this);
		}

		/// <summary>
		/// Send a call to the server
		/// </summary>
		/// <param name="type">The call type (must have been registered with Registry.RegisterClientCall)</param>
		/// <param name="data">The data pack to send</param>
		/// <param name="onReturn">The callback to be executed when the server answers this call</param>
		/// <param name="onException">The callback to be executed when the server answer this call with an exception (or the call timeouts or the connection closes)</param>
		/// <param name="timeout">The timeout (in ms), 0 means no timeout</param>
		public void SendCall(uint type, Data data, ReturnDelegate onReturn = null, ExceptionDelegate onException = null, int timeout = 60000) {
			// Validate the data
			if (!IsReady)
				throw new InvalidOperationException("The connection has already been closed");
			Registry.RegisteredCall call = Registry.GetClientCall(type);
			if (call == null)
				throw new ArgumentException("Invalid call type " + type);
			if (data.Format != call.ArgsFormat.FormatString)
				throw new ArgumentException("Invalid data type '" + data.Format + "' for call " + type);

			// Create the meta-data
			byte[] binData = data.GetBytes();
			byte[] binMeta = new Data().AddUint(type).AddUint(++LastSentId).GetBytes();
			byte[] binLength = new Data().AddUint((ulong)(binData.Length + binMeta.Length)).GetBytes();

			// Send the call
			Socket.Write(binLength);
			Socket.Write(binMeta);
			Socket.Write(binData);

			// Set timeout
			Timer interval = null;
			if (timeout != 0) {
				interval = new Timer(timeout);
				interval.AutoReset = false;
				interval.Elapsed += TimeoutCallback;
				interval.Start();
			}

			// Save info about the sent call
			PendingCalls.AddLast(new PendingCall(LastSentId, call, onReturn, onException, interval));
		}

		/// <summary>
		/// Close the conection (after all writes in the underlying socket finish)
		/// </summary>
		public void End() {
			Socket.End();
		}

		void Socket_OnConnect(object sender, EventArgs e) {
			if (OnOpen != null)
				OnOpen(this, EventArgs.Empty);
		}

		void Socket_OnClose(object sender, EventArgs e) {
			// Alert all pending calls
			LinkedListNode<PendingCall> node = PendingCalls.First;
			while (node != null) {
				if (node.Value.Interval != null)
					node.Value.Interval.Stop();
				if (node.Value.OnException != null)
					node.Value.OnException(this, -1, null);
				node = node.Next;
			}
			PendingCalls.Clear();

			Dispose();

			// Emit "close" event
			if (OnClose != null)
				OnClose(this, EventArgs.Empty);
		}

		void Socket_OnError(object sender, EventSocket.ExceptionEventArgs e) {
			try {
				if (OnError == null)
					throw e.Data;
				OnError(this, e);
			} finally {
				Dispose(true);
			}
		}

		void Socket_OnData(object sender, EventSocket.DataEventArgs args) {
			// Store the data
			Cache.Concat(args.Data);

			// Try to read messages
			while (true) {
				int length;
				BufferView backup = new BufferView(Cache);

				try {
					// Extract the size of the message
					length = (int)InflateData.ReadUint(Cache);
				} catch (Exception e) {
					if (!(e is InflateData.NotEnoughData))
						ProtocolError();
					break;
				}

				if (Cache.Length < length) {
					Cache = backup;
					break;
				}

				ProcessMessage(Cache.ExtractSlice(length));
			}
		}

		void ProcessMessage(BufferView message) {
			uint type, callId;
			// Extract the message type and sequence id
			try {
				type = (uint)InflateData.ReadUint(message);
				callId = (uint)InflateData.ReadUint(message);
			} catch {
				ProtocolError();
				return;
			}
	
			if (type != 0)
				// A call from the other side
				ProcessCall(callId, type, message);
			else {
				try {
					type = (uint)InflateData.ReadUint(message);
				} catch {
					ProtocolError();
					return;
				}
				if (type != 0)
					// An exception from the other side
					ProcessException(callId, type, message);
				else
					// A return from the other side
					ProcessReturn(callId, message);
			}
		}

		void ProcessCall(uint callId, uint type, BufferView data) {
			object inflatedData;

			// Check the sequence id
			if (callId != ++LastReceivedId) {
				ProtocolError();
				return;
			}

			// Get call definition
			Registry.RegisteredCall call = Registry.GetServerCall(type);
			if (call == null) {
				ProtocolError();
				return;
			}

			// Read the incoming data
			try {
				inflatedData = InflateData.Inflate(data, call.ArgsFormat);
			} catch {
				ProtocolError();
				return;
			}

			// Emit the "call" event
			if (OnCall != null)
				OnCall(this, new CallEventArgs(call, inflatedData, this, callId));
		}

		void ProcessException(uint callId, uint type, BufferView data) {
			PendingCall callInfo = FetchPendingCall(callId);
			if (callInfo == null) {
				// Received a timeouted (or invalid) answer
				ProtocolError();
				return;
			}
			if (!callInfo.Call.HasException(type)) {
				// Received an invalid exception type
				ProtocolError();
				return;
			}

			// Get exception definition
			Registry.RegisteredException exception = Registry.GetException(type);
			if (exception == null) {
				ProtocolError();
				return;
			}

			// Read the incoming data
			object inflatedData;
			try {
				inflatedData = InflateData.Inflate(data, exception.DataFormat);
			} catch {
				ProtocolError();
				return;
			}

			// Clear the timeout
			if (callInfo.Interval != null)
				callInfo.Interval.Stop();

			// Call the callback
			if (callInfo.OnException != null)
				callInfo.OnException(this, (int)type, inflatedData);
		}

		void ProcessReturn(uint callId, BufferView data) {
			PendingCall callInfo = FetchPendingCall(callId);
			if (callInfo == null) {
				// Received a timeouted (or invalid) answer
				ProtocolError();
				return;
			}

			// Read the incoming data
			object inflatedData;
			try {
				inflatedData = InflateData.Inflate(data, callInfo.Call.ReturnFormat);
			} catch {
				ProtocolError();
				return;
			}

			// Clear the timeout
			if (callInfo.Interval != null)
				callInfo.Interval.Stop();

			// Call the callback
			if (callInfo.OnReturn != null)
				callInfo.OnReturn(this, inflatedData);
		}

		PendingCall FetchPendingCall(uint callId) {
			LinkedListNode<PendingCall> node = PendingCalls.First;
			while (node != null) {
				if (node.Value.CallId == callId) {
					PendingCalls.Remove(node);
					return node.Value;
				}
				node = node.Next;
			}
			return null;
		}

		PendingCall FetchPendingCall(Timer interval) {
			LinkedListNode<PendingCall> node = PendingCalls.First;
			while (node != null) {
				if (node.Value.Interval == interval) {
					PendingCalls.Remove(node);
					return node.Value;
				}
				node = node.Next;
			}
			return null;
		}

		void ProtocolError() {
			End();
		}

		protected virtual void Dispose(bool disposing) {
			if (!disposed) {
				disposed = true;
				if (disposing)
					Socket.Dispose();
				Socket = null;
			}
		}

		void TimeoutCallback(object sender, ElapsedEventArgs e) {
			PendingCall callInfo = FetchPendingCall((Timer)sender);
			if (callInfo != null) {
				if (callInfo.OnException != null)
					callInfo.OnException(this, 0, null);
			}
		}

		~Connection() {
			Dispose(false);
		}
	}
}
